InfiniTAM/infinitam_ros_node.cpp

```diff
@@ -190,113 +233,286 @@ bool InfinitamNode::startInfinitam(std_srvs::SetBool::Request& request,
+    LOG(INFO) << "got point cloud";
+
+    sensor_msgs::PointCloud2 point_cloud_msg;
+    pcl::toROSMsg(*point_cloud_pcl, point_cloud_msg);
+    point_cloud_msg.header.frame_id = camera_frame_id_;
+    point_cloud_msg.header.stamp = ros::Time::now();
+
+    complete_point_cloud_pub_.publish(point_cloud_msg);
+  }
+
+  if (publish_mesh_) {
+    ROS_INFO_STREAM("publish_mesh_");
+
+    // Get the Mesh as PCL PolygonMesh .
+    extractITMMeshToPolygonMesh(triangleArray, mesh_ptr_);
+
+    ROS_INFO_STREAM("Loaded a PolygonMesh with "
+                    << mesh_ptr_->cloud.width * mesh_ptr_->cloud.height
+                    << " points and " << mesh_ptr_->polygons.size()
+                    << " polygons.");
+
+    pcl::io::saveOBJFile("../awesome_mesh.obj", *mesh_ptr_);
+
+    //     Convert PCL PolygonMesh into ROS shape_msgs Mesh.
+    convertPolygonMeshToRosMesh(mesh_ptr_, ros_scene_mesh_ptr_);
+    ROS_INFO_STREAM("Got ROS Mesh.");
+
+    // Publish ROS Mesh.
+    complete_mesh_pub_.publish(*ros_scene_mesh_ptr_);
+    ROS_INFO_STREAM("ROS Mesh published.");
+  }
+
+  if (rm_triangle_from_cuda_memory) {
+    delete cpu_triangles;
+  }
+  ROS_INFO_STREAM("publishMap end!");
+  return true;
+}
+
+void InfinitamNode::extractITMMeshToPolygonMesh(
+    const ITMMesh::Triangle* triangleArray,
+    pcl::PolygonMesh::Ptr polygon_mesh_ptr) {
+  if (triangleArray == NULL) {
+    ROS_ERROR_STREAM("triangleArray == NULL");
+    return;
+  }
+
+  std::size_t nr_triangles = 0;
+  std::size_t nr_points = 0;
+  nr_triangles = main_engine_->GetMesh()->noTotalTriangles;
+  nr_points = nr_triangles * 3;
+  ROS_INFO_STREAM("nr_triangles:  " << nr_triangles);
+  ROS_INFO_STREAM("nr_points:  " << nr_points);
+
+  pcl::PointCloud<pcl::PointXYZ>::Ptr point_cloud_pcl(
+      new pcl::PointCloud<pcl::PointXYZ>);
+
+  point_cloud_pcl->width = nr_points;
+  point_cloud_pcl->height = 1;
+  point_cloud_pcl->is_dense = true;
+  point_cloud_pcl->points.resize(point_cloud_pcl->width *
+                                 point_cloud_pcl->height);
+
+  std::size_t point_number = 0;
+
+  // All vertices of the mesh are stored in the pcl point cloud.
+  for (std::size_t i = 0; i < nr_triangles; ++i) {
+    point_cloud_pcl->points[point_number].x = triangleArray[i].p0.x;
+    point_cloud_pcl->points[point_number].y = triangleArray[i].p0.y;
+    point_cloud_pcl->points[point_number].z = triangleArray[i].p0.z;
+    ++point_number;
+    point_cloud_pcl->points[point_number].x = triangleArray[i].p1.x;
+    point_cloud_pcl->points[point_number].y = triangleArray[i].p1.y;
+    point_cloud_pcl->points[point_number].z = triangleArray[i].p1.z;
+    ++point_number;
+    point_cloud_pcl->points[point_number].x = triangleArray[i].p2.x;
+    point_cloud_pcl->points[point_number].y = triangleArray[i].p2.y;
+    point_cloud_pcl->points[point_number].z = triangleArray[i].p2.z;
+    ++point_number;
+  }
+
+  // Build the point cloud.
+  pcl::toROSMsg(*point_cloud_pcl, mesh_ptr_->cloud);
```

Comment:
Why don't you use here polygon_mesh_ptr instead of mesh_ptr_ ? Replace all mesh_ptr_ in this method with polygon_mesh_ptr.